### PR TITLE
[IOTDB-5454] Support shuffle function of DataExchangeModule

### PR DESCRIPTION
**Pull request by lancelly**
related desgin doc:
https://apache-iotdb.feishu.cn/docx/QOsOdydlBo0cBVxZLeycVulgnAb
https://apache-iotdb.feishu.cn/docx/Hoimd9Fo5ok4rhxa1shchTs9nqV
https://apache-iotdb.feishu.cn/docx/IJHOdjN0voIMvIxsAKFc8LxJnRc

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java [hunk 1]
```diff
@@ -468,19 +483,18 @@ public synchronized ISinkHandle createLocalSinkHandleForFragment(
    */
   public ISinkHandle createLocalSinkHandleForPipeline(
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java:484]:
```suggestion
  public ISinkChannel createLocalSinkChannelForPipeline(
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java [hunk 2]
```diff
@@ -419,20 +442,15 @@ public MPPDataExchangeServiceImpl getOrCreateMPPDataExchangeServiceImpl() {
     return mppDataExchangeService;
   }
 
-  @Override
   public synchronized ISinkHandle createLocalSinkHandleForFragment(
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java]:
```suggestion
  public synchronized ISinkChannel createLocalSinkChannelForFragment(
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java [hunk 3]
```diff
@@ -468,19 +483,18 @@ public synchronized ISinkHandle createLocalSinkHandleForFragment(
...
             localMemoryManager);
     return new LocalSinkHandle(
         queue,
-        new PipelineSinkHandleListenerImpl(
+        new SinkHandleListenerImpl(
             driverContext.getFragmentInstanceContext(), driverContext::failed));
   }
 
-  @Override
   public ISinkHandle createSinkHandle(
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java]:
```suggestion
  public ISinkChannel createSinkChannel(
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java [hunk 4]
```diff
@@ -490,30 +504,84 @@ public ISinkHandle createSinkHandle(
...
+      String localPlanNodeId,
+      // TODO: replace with callbacks to decouple MPPDataExchangeManager from
+      // FragmentInstanceContext
+      FragmentInstanceContext instanceContext) {
+    if (shuffleSinkHandles.containsKey(localFragmentInstanceId)) {
+      throw new IllegalStateException(
+          "ShuffleSinkHandle for " + localFragmentInstanceId + " exists.");
+    }
+
+    List<ISinkHandle> downStreamChannelList =
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java:542]:
```suggestion
    List<ISinkChannel> downStreamChannelList =
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java [hunk 5]
```diff
@@ -490,30 +504,84 @@ public ISinkHandle createSinkHandle(
...
+        downStreamChannelLocationList.stream()
+            .map(
+                downStreamChannelLocation ->
+                    createHandleForShuffleSink(
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java:546]:
```suggestion
                    createChannelForShuffleSink(
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java [hunk 6]
```diff
@@ -490,30 +504,84 @@ public ISinkHandle createSinkHandle(
...
+            downStreamChannelList,
+            downStreamChannelIndex,
+            shuffleStrategyEnum,
+            localPlanNodeId,
+            new ShuffleSinkHandleListenerImpl(instanceContext, instanceContext::failed));
+    shuffleSinkHandles.put(localFragmentInstanceId, shuffleSinkHandle);
+    return shuffleSinkHandle;
+  }
+
+  private ISinkHandle createHandleForShuffleSink(
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/MPPDataExchangeManager.java:565]:
```suggestion
  private ISinkChannel createChannelForShuffleSink(
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/IMPPDataExchangeManager.java [hunk 7]
```diff
@@ -20,36 +20,35 @@
...
       String localPlanNodeId,
       FragmentInstanceContext instanceContext);
 
-  ISinkHandle createLocalSinkHandleForFragment(
-      TFragmentInstanceId localFragmentInstanceId,
-      TFragmentInstanceId remoteFragmentInstanceId,
-      String remotePlanNodeId,
-      FragmentInstanceContext instanceContext);
-
+  ISinkHandle createLocalSinkHandleForPipeline(DriverContext driverContext, String planNodeId);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/IMPPDataExchangeManager.java:51]:
```suggestion
  ISinkChannel createLocalSinkChannelForPipeline(DriverContext driverContext, String planNodeId);
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/sink/SinkHandle.java [hunk 8]
```diff
@@ -57,9 +57,9 @@
 import static org.apache.iotdb.db.mpp.metric.DataExchangeCountMetricSet.SEND_NEW_DATA_BLOCK_NUM_CALLER;
 import static org.apache.iotdb.tsfile.read.common.block.TsBlockBuilderStatus.DEFAULT_MAX_TSBLOCK_SIZE_IN_BYTES;
 
-public class SinkHandle implements ISinkHandle {
+public class SinkHandle implements ISinkHandle, ISinkChannel {
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/exchange/sink/SinkHandle.java:60]:
```suggestion
public class SinkChannel implements ISinkChannel {
```

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/sink/ShuffleHelperOperator.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (children.get(downStreamChannelIndex.getCurrentIndex()).hasNext()) {
+      return true;
+    }
+    int currentIndex = downStreamChannelIndex.getCurrentIndex();
+    // current channel have no more data
+    sinkHandle.setNoMoreTsBlocksOfOneChannel(downStreamChannelIndex.getCurrentIndex());
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/sink/ShuffleHelperOperator.java:68]:
```suggestion
    int currentIndex = downStreamChannelIndex.getCurrentIndex();
    if (children.get(currentIndex).hasNext()) {
      return true;
    }
    // current channel have no more data
    sinkHandle.setNoMoreTsBlocksOfOneChannel(currentIndex);
```

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/OperatorTreeGenerator.java [hunk 10]
```diff
@@ -1822,36 +1829,71 @@ public Operator visitExchange(ExchangeNode node, LocalExecutionPlanContext conte
...
+  public Operator visitIdentitySink(IdentitySinkNode node, LocalExecutionPlanContext context) {
+    context.addExchangeSumNum(1);
+    OperatorContext operatorContext =
+        context
+            .getDriverContext()
+            .addOperatorContext(
+                context.getNextOperatorId(),
+                node.getPlanNodeId(),
+                IdentitySinkOperator.class.getSimpleName());
+    context.getTimeSliceAllocator().recordExecutionWeight(operatorContext, 1);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/OperatorTreeGenerator.java:1841]:
```suggestion
```
no need to record it, it's not a time consuming operation.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/OperatorTreeGenerator.java [hunk 11]
```diff
@@ -1822,36 +1829,71 @@ public Operator visitExchange(ExchangeNode node, LocalExecutionPlanContext conte
   }
 
   @Override
-  public Operator visitFragmentSink(FragmentSinkNode node, LocalExecutionPlanContext context) {
-    if (!isSameNode(node.getDownStreamEndpoint())) {
-      context.addExchangeSumNum(1);
-    }
-    Operator child = node.getChild().accept(this, context);
+  public Operator visitIdentitySink(IdentitySinkNode node, LocalExecutionPlanContext context) {
+    context.addExchangeSumNum(1);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/OperatorTreeGenerator.java:1833]: why we don't need the previous same node judgement?

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/OperatorTreeGenerator.java [hunk 12]
```diff
@@ -1822,36 +1829,71 @@ public Operator visitExchange(ExchangeNode node, LocalExecutionPlanContext conte
...
+            context.getInstanceContext());
+    sinkHandle.setMaxBytesCanReserve(context.getMaxBytesOneHandleCanReserve());
+    context.getDriverContext().setSinkHandle(sinkHandle);
+
+    return new IdentitySinkOperator(operatorContext, children, downStreamChannelIndex, sinkHandle);
+  }
+
+  @Override
+  public Operator visitShuffleSink(ShuffleSinkNode node, LocalExecutionPlanContext context) {
+    context.addExchangeSumNum(1);
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/OperatorTreeGenerator.java:1867]: same as above

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/test/java/org/apache/iotdb/db/mpp/plan/plan/node/sink/IdentitySinkNodeSerdeTest.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        new DownStreamChannelLocation(
+            new TEndPoint("test", 1), new TFragmentInstanceId("test", 1, "test"), "test");
+    IdentitySinkNode identitySinkNode1 =
+        new IdentitySinkNode(
+            new PlanNodeId("testIdentitySinkNode"),
+            Collections.singletonList(downStreamChannelLocation));
+    ByteBuffer byteBuffer = ByteBuffer.allocate(1024);
+    identitySinkNode1.serialize(byteBuffer);
+    byteBuffer.flip();
+    assertEquals(PlanNodeDeserializeHelper.deserialize(byteBuffer), identitySinkNode1);
```
Comment (SzyWilliam) [inline server/src/test/java/org/apache/iotdb/db/mpp/plan/plan/node/sink/IdentitySinkNodeSerdeTest.java:51]: we shall put the expect value as the 1st param.

**Review by SzyWilliam — APPROVED** [review]
bravo👍🏻

File: server/src/test/java/org/apache/iotdb/db/mpp/plan/plan/node/sink/IdentitySinkNodeSerdeTest.java (hunk 13 above)
Comment (lancelly) [reply server/src/test/java/org/apache/iotdb/db/mpp/plan/plan/node/sink/IdentitySinkNodeSerdeTest.java]: Fixed. Excellent advice 0.o

**Pull request merged**